tests/Unit/EmogrifierTest.php

```diff
@@ -684,6 +684,92 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+            'id more specific than 20 classes and types'
+                => ['#p4', 'html body ' . str_repeat('.p-4', 20), '>', '<p class="p-4" id="p4"'],
+            'pseudo-class as specific as class' => ['.p-1', '*:first-child', '==', '<p class="p-1"'],
+            'attribute as specific as class' => ['span[title="bonjour"]', '.p-2 span', '==', '<span title="bonjour"'],
+            ':not alone does not increase specificity' => ['p:not(* + *)', 'p', '==', '<p class="p-1"'],
+            ':not with type more specific than nothing' => ['.p-1:not(h1)', '.p-1', '>', '<p class="p-1"'],
+            ':not with type as specific as type' => ['*:not(h1)', 'p', '==', '<p class="p-1"'],
+            ':not with several types more specific than single type'
+                => ['*:not(html body h1)', 'p', '>', '<p class="p-1"'],
+            ':not with class more specific than types'
+                => ['*:not(.p-2)', 'html body p', '>', '<p class="p-1"'],
+            ':not with class more specific than :not with types'
+                => ['*:not(.p-2)', '*:not(html body h1)', '>', '<p class="p-1"'],
+            'class more specific than :not with types' => ['.p-1', '*:not(html body h1)', '>', '<p class="p-1"'],
+        ];
+
+        /**
+         * @var string[][] $parameterSets Sets of parameters returned by this data provider for
+         *      {@see EmogrifierTest::emogrifyAppliesMostSpecificSelectorCssToMatchingElements}.
+         *      The sprintf placeholders %1$s and %2$s will automatically be replaced with CSS declarations
+         *      like 'color: red;' or 'color: blue;'.
+         */
+        $parameterSets = [];
+        foreach ($selectorData as $name => $data) {
+            $firstRuleBlock = $data[0] . ' { %1$s }';
+            $secondRuleBlock = $data[1] . ' { %2$s }';
+            $firstRuleBlockWasApplied = $data[3] . ' style="%1$s">';
+            $secondRuleBlockWasApplied = $data[3] . ' style="%2$s">';
+
+            // Check selectors actually match on their own (i.e. test is valid)
+            $parameterSets[$name . ' (1st selector valid for test)'] = [$firstRuleBlock, $firstRuleBlockWasApplied];
+            $parameterSets[$name . ' (2nd selector valid for test)'] = [$secondRuleBlock, $secondRuleBlockWasApplied];
+
+            // Check first selector matches when its rule is last unless the second has higher specificity
+            $parameterSets[$name . ' (1st selector in last rule)'] = [
+                $secondRuleBlock . ' ' . $firstRuleBlock,
+                $data[2] === '<' ? $secondRuleBlockWasApplied : $firstRuleBlockWasApplied
+            ];
+
+            // Check second selector matches when its rule is last unless the first has higher specificity
+            $parameterSets[$name . ' (2nd selector in last rule)'] = [
+                $firstRuleBlock . ' ' . $secondRuleBlock,
+                $data[2] === '>' ? $firstRuleBlockWasApplied : $secondRuleBlockWasApplied
+            ];
```

Comment:
Instead of having code blocks with comments explaining what they do, please move these code blocks to private methods with "speaking" names. This will make it easier to get an overview of what happens here.

Reply:
Refactored into new private methods `getMatchedCssDataSet` and `getSelectorSpecificityCssDataSet`